security/rbac/engine/engine.go

```diff
@@ -0,0 +1,197 @@
+type rbacEngine struct {
+	action     pb.RBAC_Action
+	conditions map[string]*cel.Program
+}
+
+// Creates a new rbacEngine from an RBAC.
+func newRbacEngine(rbac pb.RBAC) rbacEngine {
+	action := rbac.Action
+	conditions := make(map[string]*cel.Program)
+	for policyName, policy := range rbac.Policies {
+		conditions[policyName] = exprToProgram(policy.Condition)
+	}
+	return rbacEngine{action, conditions}
+}
+
+// Returns whether a set of authorization arguments, as contained in evalMap, match an
+//  rbacEngine's conditions.
+// If any policy matches, the engine has been matched, and the list of matching policy
+//  names will be returned.
+// Else if any policy is missing attributes, the engine is not matched, and the list of
+//  policy names that can't be evaluated due to missing attributes will be returned.
+// Else, there is not a match.
+func (engine rbacEngine) matches(evalMap map[string]interface{}) (bool, []string) {
+	matchingPolicyNames := []string{}
+	unknownPolicyNames := []string{}
+	for policyName, condition := range engine.conditions {
+		match, err := matchesCondition(condition, evalMap)
+		if err != nil {
+			unknownPolicyNames = append(unknownPolicyNames, policyName)
+		} else if match {
+			matchingPolicyNames = append(matchingPolicyNames, policyName)
+		}
+	}
+	if len(matchingPolicyNames) > 0 {
+		return true, matchingPolicyNames
+	}
+	return false, unknownPolicyNames
+}
+
+// CelEvaluationEngine is the struct for CEL engine.
+type CelEvaluationEngine struct {
+	engines []rbacEngine
+}
+
+// NewCelEvaluationEngine builds a cel evaluation engine from a list of Envoy RBACs.
+func NewCelEvaluationEngine(rbacs []pb.RBAC) (CelEvaluationEngine, error) {
+	if len(rbacs) < 1 || len(rbacs) > 2 {
+		return CelEvaluationEngine{}, status.Errorf(codes.InvalidArgument, "must provide 1 or 2 RBACs")
+	}
+	if len(rbacs) == 2 && (rbacs[0].Action != pb.RBAC_DENY || rbacs[1].Action != pb.RBAC_ALLOW) {
+		return CelEvaluationEngine{}, status.Errorf(codes.InvalidArgument, "when providing 2 RBACs, must have 1 DENY and 1 ALLOW in that order")
+	}
+	var engines []rbacEngine
+	for _, rbac := range rbacs {
+		engines = append(engines, newRbacEngine(rbac))
+	}
+	return CelEvaluationEngine{engines}, nil
+}
+
+// Evaluate is the core function that evaluates whether an RPC is authorized.
+//
+// ALLOW policy. If one of the RBAC conditions is evaluated as true, then the
+// CEL engine evaluation returns allow. If all of the RBAC conditions are
+// evaluated as false, then it returns deny. Otherwise, some conditions are
+// false and some are unknown, it returns undecided.
+//
+// DENY policy. If one of the RBAC conditions is evaluated as true, then the
+// CEL engine evaluation returns deny. If all of the RBAC conditions are
+// evaluated as false, then it returns allow. Otherwise, some conditions are
+// false and some are unknown, it returns undecided.
+//
+// DENY policy + ALLOW policy. If one of the conditions in the DENY policy is
```

Comment:
This rule is updated as: 

*DENY policy + ALLOW policy. Evaluation is in the following order: If one of the expressions in the DENY policy is true, the authorization engine returns deny. If one of the expressions in the DENY policy is unknown, it returns undecided. Now all the expressions in the DENY policy are false, it returns the evaluation of the ALLOW policy.*

Can you please check:
1. Are there any differences in what we are doing and this new update? (I think there are not, but can you please double-check this?)
2. If answer to 1 is no, can we modify our code logic to the description above? I think generally it's a better representation of the combination rules.  
Thanks!

Reply:
1. Yes, currently if one of the expressions in the DENY policy is unknown, it returns the evaluation of the ALLOW policy. Undecided is only returned if both policies are completely unmatched and something is unknown.
2. Yes, I'll change it. Thanks!

Reply:
Elizabeth, I changed the evaluate sequence on DENY+ALLOW policies recently based on suggestion from @ejona86. If there is an expression in DENY policy is unknown, we should not allow it (even if there is an expression in ALLOW policy evaluated as true). 

I know the design is a moving target. But we should sync up this design change.

Reply:
Got it, no worries!